js/lib/src/instructions/matches.ts

```diff
@@ -0,0 +1,488 @@
+  getOracle,
+  getPlayerPDA,
+} from "../utils/pda";
+
+export interface CreateMatchArgs {
+  matchState: AnchorMatchState;
+  tokenEntryValidationRoot: null;
+  tokenEntryValidation: null | AnchorTokenEntryValidation[];
+  winOracle: web3.PublicKey;
+  winOracleCooldown: BN;
+  authority: web3.PublicKey;
+  space: BN;
+  leaveAllowed: boolean;
+  joinAllowedDuringStart: boolean;
+  minimumAllowedEntryTime: BN | null;
+  desiredNamespaceArraySize?: BN;
+}
+
+export interface UpdateMatchArgs {
+  matchState: AnchorMatchState;
+  tokenEntryValidationRoot: null;
+  tokenEntryValidation: null;
+  winOracleCooldown: BN;
+  authority: web3.PublicKey;
+  leaveAllowed: boolean;
+  joinAllowedDuringStart: boolean;
+  minimumAllowedEntryTime: BN | null;
+}
+
+export interface JoinMatchArgs {
+  amount: BN;
+  tokenEntryValidationProof: null;
+  tokenEntryValidation: null;
+}
+
+export interface LeaveMatchArgs {
+  amount: BN;
+}
+
+export interface DisburseTokensByOracleArgs {
+  tokenDeltaProofInfo: null;
+}
+
+export interface CreateMatchAdditionalArgs {
+  seed: string;
+  finalized: boolean;
+  tokenTransferRoot: null;
+  tokenTransfers: null | AnchorTokenDelta[];
+}
+
+export interface CreateOrUpdateOracleArgs {
+  seed: string;
+  authority: web3.PublicKey;
+  space: BN;
+  finalized: boolean;
+  tokenTransferRoot: null;
+  tokenTransfers: AnchorTokenDelta[] | null;
+}
+
+export interface DrainMatchArgs {}
+
+export interface DrainOracleArgs {
+  seed: string;
+  authority: web3.PublicKey;
+}
+
+export interface UpdateMatchFromOracleAccounts {
+  winOracle: web3.PublicKey;
+}
+
+export interface UpdateMatchAccounts {
+  winOracle: web3.PublicKey;
+}
+
+export interface DrainMatchAccounts {
+  receiver: web3.PublicKey | null;
+}
+
+export interface DrainOracleAccounts {
+  receiver: web3.PublicKey | null;
+}
+
+export interface DisburseTokensByOracleAccounts {
+  winOracle: web3.PublicKey;
+}
+
+export interface JoinMatchAccounts {
+  tokenMint: web3.PublicKey;
+  sourceTokenAccount: web3.PublicKey | null;
+  tokenTransferAuthority: web3.Keypair | null;
+  validationProgram: web3.PublicKey | null;
+}
+
+export interface LeaveMatchAccounts {
+  tokenMint: web3.PublicKey;
+  receiver: web3.PublicKey;
+}
+
+export interface JoinMatchAdditionalArgs {
+  sourceType: TokenType;
+  index: BN | null;
+  winOracle: web3.PublicKey;
+}
+
+export interface LeaveMatchAdditionalArgs {
+  winOracle: web3.PublicKey;
+}
+
+export interface DrainMatchAdditionalArgs {
+  winOracle: web3.PublicKey;
+}
+
+export interface DisburseTokensByOracleAdditionalArgs {
+  tokenDelta: AnchorTokenDelta;
+}
+
+export class Instruction extends SolKitInstruction {
+  constructor(args: { program: Program.Program }) {
+    super(args);
+  }
+  async createMatch(
+    args: CreateMatchArgs,
+    _accounts = {},
+    _additionalArgs = {}
+  ) {
+    const [match, _matchBump] = await getMatch(args.winOracle);
+
+    transformTokenValidations(args);
+    return {
```

Comment:
I like this return type format. We should formalize it into sol-kit and update the rest of raindrops at some point.